tests/c/tcp-ping-pong.c

```diff
@@ -0,0 +1,346 @@
+ * server()                                                                                                           *
+ *====================================================================================================================*/
+
+/**
+ * @brief TCP echo server.
+ *
+ * @param argc  Argument count.
+ * @param argv  Argument list.
+ * @param local Local socket address.
+ */
+static void server(int argc, char *const argv[], struct sockaddr_in *local)
+{
+    int qd = -1;
+    int nbytes = 0;
+    int sockqd = -1;
+
+    /* Initialize demikernel */
+    assert(demi_init(argc, argv) == 0);
+
+    /* Setup local socket. */
+    assert(demi_socket(&sockqd, AF_INET, SOCK_STREAM, 0) == 0);
+    assert(demi_bind(sockqd, (const struct sockaddr *)local, sizeof(struct sockaddr_in)) == 0);
+    assert(demi_listen(sockqd, 16) == 0);
+
+    /* Accept client . */
+    qd = accept_wait(sockqd);
+
+    /* Run. */
+    while (nbytes < MAX_BYTES)
+    {
+        demi_qresult_t qr = {0};
+        demi_sgarray_t sga = {0};
+
+        /* Pop scatter-gather array. */
+        pop_wait(qd, &qr);
+
+        /* Extract received scatter-gather array. */
+        memcpy(&sga, &qr.qr_value.sga, sizeof(demi_sgarray_t));
+
+        nbytes += sga.sga_segs[0].sgaseg_len;
+
+        /* Push scatter-gather array. */
+        push_wait(qd, &sga, &qr);
+
+        /* Release received scatter-gather array. */
+        assert(demi_sgafree(&sga) == 0);
+
+        fprintf(stdout, "ping (%d)\n", nbytes);
+    }
+}
+
+/*====================================================================================================================*
+ * client()                                                                                                           *
+ *====================================================================================================================*/
+
+/**
+ * @brief TCP echo client.
+ *
+ * @param argc   Argument count.
+ * @param argv   Argument list.
+ * @param remote Remote socket address.
+ */
+static void client(int argc, char *const argv[], const struct sockaddr_in *remote)
+{
+    int nbytes = 0;
+    int sockqd = -1;
+
+    /* Initialize demikernel */
+    assert(demi_init(argc, argv) == 0);
+
+    /* Setup socket. */
+    assert(demi_socket(&sockqd, AF_INET, SOCK_STREAM, 0) == 0);
+
+    /* Connect to server. */
+    connect_wait(sockqd, remote);
+
+    /* Run. */
+    while (nbytes < MAX_BYTES)
+    {
+        demi_qresult_t qr = {0};
+        demi_sgarray_t sga = {0};
+
+        /* Allocate scatter-gather array. */
+        sga = demi_sgaalloc(DATA_SIZE);
+        assert(sga.sga_segs != 0);
+
+        /* Cook data. */
+        memset(sga.sga_segs[0].sgaseg_buf, 1, DATA_SIZE);
+
+        /* Push scatter-gather array. */
+        push_wait(sockqd, &sga, &qr);
+
+        /* Release sent scatter-gather array. */
+        assert(demi_sgafree(&sga) == 0);
+
+        /* Pop data scatter-gather array. */
+        pop_wait(sockqd, &qr);
```

Comment:
Shouldn't we check the result for something?  (Beyond what pop_wait() does internally, I mean)

Note that since TCP doesn't guarantee preservation of message boundaries, there is no guarantee here that we'll get back the same number of bytes that we sent in the push_wait above.  But you could turn this into a loop of pop_wait()s that continue until the expected number of bytes come back.

Reply:
Sure. I missed this. I introduced a simple payload checking logic, similarly to what we have in the rust test.